src/python/torchdistx/gossip_grad.py

```diff
@@ -0,0 +1,338 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+# All rights reserved.
+#
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree.
+
+import math
+import random
+from enum import Enum, auto
+from itertools import cycle
+
+import torch
+import torch.distributed as dist
+import torch.distributed.distributed_c10d as c10d
+from torch._C._distributed_c10d import ProcessGroup
+from torch.distributed.algorithms._comm_hooks import default
+
+
+class Topology(Enum):
+    r"""
+    Specifies which topology will be used as a base for gradient communication.
+    For more information, please refer to the original
+    `paper <https://arxiv.org/abs/1803.05880>`_
+
+    CUBE: A hypercube topology - a hierarchical virtual organization of compute nodes.
+        For this topology gossiping is happanning with a neighbouring vertex.
+              *----*
+             /|   /|
+            *----* |
+            | * -|-*
+            |/   |/
+            *----*
+
+    DISSEMINATION: A dissemination topology has similar property
+        as hypercube virtual topology.
+        For this topology gossiping is happanning with the neighbouring node,
+        then every 2nd node, every 4th, etc.
+
+            .  *  .
+          *          *
+        .              .
+        *              *
+        .              .
+          *          *
+            .  *  .
+
+    .. note::
+        Current implementation does not support uneven number of nodes for a CUBE
+        topology.
+
+    """
+    CUBE = auto()
```

Comment:
Could we add documentation on which one is advised to use initially, and maybe start thinking about various heuristics / tuning that we could do automatically determine the best topology?

Reply:
@rohan-varma sure, I mentioned this in a main thread. I would like to experiment first and then develop a more clear heuristic